### DISCUSSION TITLE
parinfer

### DISCUSSION THREAD

**Discussion by Suya1671** [helix-editor/helix#4090] [Ideas] [answered]
when working with lisps, https://github.com/eraserhd/parinfer-rust is a great tool to make the experience much better. Maybe there could be a way to directly interface with parinfer-rust since its well, rust and has a `lib.rs` file. Although it doesn't seem to reexport many functions.

**Comment by sudormrfbin** [comment, marked as answer]
If it is decided that the plugin scripting language will be [some variant of lisp][1], native integration with parinfer would be nice to have. Otherwise a plugin would be the appropriate path. Either way a plugin system will have to be put in place first.

[1]: https://github.com/helix-editor/helix/pull/2949#issuecomment-1179686877

**Comment by chopfitzroy** [comment]
Adding to this conversation in the event the plugin system does end up being a lisp variant.

Another option to consider when implementing s-expression structural editing might be something similar to [`guns/vim-sexp`](https://github.com/guns/vim-sexp) but with bindings similar to those provided by [`tpope/vim-sexp-mappings-for-regular-people`](https://github.com/tpope/vim-sexp-mappings-for-regular-people/) as they potentially align better with the Helix defaults.

Another potential project to take inspiration from is [`abo-abo/lispy`](https://github.com/abo-abo/lispy/) which is an Emacs plugin that uses _vim-like_ key bindings.

Some other useful notes on `parinfer` and s-expression structural editing in general can be found at:
- [Parinfer – Simpler Lisp Editing (HN)](https://news.ycombinator.com/item?id=17633055)
- [`noctuid/parinfer-notes`](https://github.com/noctuid/parinfer-notes)
